source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst

```diff
@@ -455,18 +503,33 @@ I/O.
 Checking Out a Connection
 -------------------------
 
-A Pool MUST have a method of allowing the driver to check out a `Connection <#connection>`_. Checking out a `Connection <#connection>`_ involves entering the WaitQueue and waiting for a `Connection <#connection>`_ to become available. If the thread times out in the WaitQueue, an error is thrown.
-
-If, in the process of iterating available `Connections <#connection>`_ in the
-pool by the checkOut method, a perished `Connection <#connection>`_ is
-encountered, such a `Connection <#connection>`_ MUST be closed (as described in
-`Closing a Connection <#closing-a-connection-internal-implementation>`_) and the
-iteration of available `Connections <#connection>`_ MUST continue until either a
+A Pool MUST have a method of allowing the driver to check out a `Connection
+<#connection>`_. Checking out a `Connection <#connection>`_ involves entering
+the WaitQueue, waiting to be granted access to the list of available
+connections, and finding or creating a `Connection <#connection>`_ to be
+returned. If the thread times out in the WaitQueue, an error is thrown.
+
+Once reaching the front of the WaitQueue, a thread begins iterating over the
+list of available `Connections <#connection>`_, searching for a non-perished one
+to be returned. If a perished `Connection <#connection>`_ is encountered, such a
+`Connection <#connection>`_ MUST be closed (as described in `Closing a
+Connection <#closing-a-connection-internal-implementation>`_) and the iteration
+of available `Connections <#connection>`_ MUST continue until either a
 non-perished available `Connection <#connection>`_ is found or the list of
-available `Connections <#connection>`_ is exhausted. If no `Connections
-<#connection>`_ are available and the total number of `Connections
-<#connection>`_ is less than maxPoolSize, the pool MUST create a `Connection
-<#connection>`_, establish it, mark it as "in use" and return it.
+available `Connections <#connection>`_ is exhausted.
+
+If the list is exhausted, the total number of `Connections <#connection>`_ is
+less than maxPoolSize, and pendingConnectionCount < maxConnecting, the pool MUST
+create a `Connection <#connection>`_, establish it, mark it as "in use" and
+return it. If totalConnectionCount == maxPoolSize or pendingConnectionCount ==
+maxConnecting, then the thread MUST wait until either both of those conditions
+are met or until a `Connection <#connection>`_ becomes available, re-entering
+the checkOut loop once it finishes waiting. This waiting MUST NOT block other
+threads from checking in `Connections <#connection>`_ to the pool. Threads that
+are waiting MUST be notified in order that they entered the WaitQueue. For
+drivers that implement the WaitQueue via a fair semaphore, a second semaphore
```

Comment:
Using two semaphores with two distinct waiting steps can lead to unfair behavior. For example,

1. Thread A enters the wait queue 
1. Thread B enters the wait queue 
1. Thread A needs to create a connection but pendingConnectionCount == maxConnecting.
1. Another thread finishes an operation and checks a connection back into the pool.
1. Thread B is unblocked and checks out the connection before Thread A.
1. Another thread finishes creating its connection and decrements pendingConnectionCount.
1. Thread A is unblocked and creates a new connection (or grabs an available one) after Thread B.

Is that acceptable? I would argue yes and contend that strict fairness should not be a major concern. My reasoning is that the order the threads enter the wait queue is already non-deterministic. Actually this behavior exists even without maxConnecting: during the time it takes to create a new connection, an existing connection may be checked back into the pool which will unblock a "later" thread.